src/physical/tabular/operations/triescan_prune.rs

```diff
@@ -0,0 +1,473 @@
+                }
+            } else if self.input_trie_scan_current_layer
+                == self.input_trie_scan.get_types().len() - 1
+            {
+                // Lowest layer has been reached and an materialized value has been found
+                return true;
+            } else {
+                self.down();
+            }
+        }
+    }
+}
+
+impl<'a> TrieScanPrune<'a> {
+    /// Construct new [`TrieScanPrune`] object.
+    pub fn new(input_trie_scan: TrieScanEnum<'a>) -> Self {
+        let target_types = input_trie_scan.get_types().clone();
+
+        let mut output_column_scans: Vec<UnsafeCell<ColumnScanT<'a>>> = Vec::new();
+
+        let state = Rc::new(UnsafeCell::new(TrieScanPruneState {
+            input_trie_scan,
+            input_trie_scan_current_layer: 0,
+            column_peeks: vec![false; target_types.len()],
+            initialized: false,
+            external_current_layer: 0,
+        }));
+
+        // Create one `ColumnScanPrune` for every input column
+        for (i, target_type) in target_types.iter().enumerate() {
+            // Generate code for every possible data type of the input column
+            macro_rules! create_column_scan_for_datatype {
+                ($variant:ident, $type:ty) => {{
+                    // Get input column scan
+                    // SAFETY: we're the only one accessing the shared state at this moment
+                    let column_scan_cell = unsafe { (*state.get()).input_trie_scan.get_scan(i) };
+
+                    let column_scan_cell = column_scan_cell.unwrap();
+
+                    // Create output column scan
+                    if let ColumnScanT::$variant(referenced_scan_cell) =
+                        unsafe { &*column_scan_cell.get() }
+                    {
+                        output_column_scans.push(UnsafeCell::new(ColumnScanT::$variant(
+                            ColumnScanCell::new(ColumnScanEnum::ColumnScanPrune(
+                                ColumnScanPrune::new(Rc::clone(&state), i, referenced_scan_cell),
+                            )),
+                        )));
+                    } else {
+                        panic!("Expected a column scan of type {}", stringify!($type));
+                    }
+                }};
+            }
+
+            match target_type {
+                StorageTypeName::U32 => create_column_scan_for_datatype!(U32, u32),
+                StorageTypeName::U64 => create_column_scan_for_datatype!(U64, u64),
+                StorageTypeName::Float => create_column_scan_for_datatype!(Float, Float),
+                StorageTypeName::Double => create_column_scan_for_datatype!(Double, Double),
+            };
+        }
+
+        Self {
+            state,
+            output_column_scans,
+            target_types,
+        }
+    }
+}
+
+impl<'a> TrieScan<'a> for TrieScanPrune<'a> {
+    fn up(&mut self) {
+        unsafe {
+            (*self.state.get()).external_up();
```

Comment:
Again, `get_mut` might help getting rid of the `unsafe` block. (Same in methods below.)